File: example/src/main/scala/com/github/j5ik2o/pekko/persistence/effector/example/scalaimpl/persistenceEffector/fp/BankAccountAggregate.scala
package com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.persistenceEffector.fp

import com.github.j5ik2o.pekko.persistence.effector.example.*
import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.*
import com.github.j5ik2o.pekko.persistence.effector.example.scalaimpl.persistenceEffector.{
  BankAccount,
  BankAccountAggregateState,
}
import com.github.j5ik2o.pekko.persistence.effector.scaladsl.*
import org.apache.pekko.actor.typed.Behavior
import org.apache.pekko.actor.typed.scaladsl.Behaviors

object BankAccountAggregate {
  def actorName(aggregateId: BankAccountId): String =
    s"${aggregateId.aggregateTypeName}-${aggregateId.asString}"

  def apply(
    aggregateId: BankAccountId,
    persistenceMode: PersistenceMode = PersistenceMode.Persisted,
  ): Behavior[BankAccountCommand] = {
    val config = PersistenceEffectorConfig
      .create[BankAccountAggregateState, BankAccountEvent, BankAccountCommand](
        persistenceId = PersistenceId.ofUniqueId(actorName(aggregateId)),
        initialState = BankAccountAggregateState.NotCreated(aggregateId),
        applyEvent = (state, event) => state.applyEvent(event),
      )
      .withPersistenceMode(persistenceMode)
      .withSnapshotCriteria(SnapshotCriteria.every(2))
      .withRetentionCriteria(RetentionCriteria.snapshotEvery(2))
    Behaviors.setup[BankAccountCommand] { implicit ctx =>
      PersistenceEffector
        .fromConfig[BankAccountAggregateState, BankAccountEvent, BankAccountCommand](
          config,
        ) {
          case (initialState: BankAccountAggregateState.NotCreated, effector) =>
            handleNotCreated(initialState, effector)
          case (initialState: BankAccountAggregateState.Created, effector) =>
            handleCreated(initialState, effector)
        }
    }
  }

  private def handleNotCreated(
    state: BankAccountAggregateState.NotCreated,
    effector: PersistenceEffector[BankAccountAggregateState, BankAccountEvent, BankAccountCommand])
    : Behavior[BankAccountCommand] =
    Behaviors.receiveMessagePartial { case BankAccountCommand.Create(aggregateId, limit, balance, replyTo) =>
      val Result(bankAccount, event) = BankAccount.create(aggregateId, limit, balance)
      effector.persistEvent(event) { _ =>
        replyTo ! CreateReply.Succeeded(aggregateId)
        val newState: BankAccountAggregateState.Created =
          BankAccountAggregateState.Created(state.aggregateId, bankAccount)
        handleCreated(newState, effector)
      }
    }

  private def handleCreated(
    state: BankAccountAggregateState.Created,
    effector: PersistenceEffector[BankAccountAggregateState, BankAccountEvent, BankAccountCommand])
    : Behavior[BankAccountCommand] =
    effector.persistSnapshot(state) { _ =>
      Behaviors.receiveMessagePartial {
        case BankAccountCommand.Stop(aggregateId, replyTo) =>
          replyTo ! StopReply.Succeeded(aggregateId)
          Behaviors.stopped
        case BankAccountCommand.GetBalance(aggregateId, replyTo) =>
          replyTo ! GetBalanceReply.Succeeded(aggregateId, state.bankAccount.balance)
          Behaviors.same
        case BankAccountCommand.DepositCash(aggregateId, amount, replyTo) =>
          state.bankAccount
            .add(amount)
            .fold(
              error => {
                replyTo ! DepositCashReply.Failed(aggregateId, error)
                Behaviors.same
              },
              { case Result(newBankAccount, event) =>
                effector.persistEvent(event) { _ =>
                  replyTo ! DepositCashReply.Succeeded(aggregateId, amount)
                  handleCreated(state.copy(bankAccount = newBankAccount), effector)
                }
              },
            )
        case BankAccountCommand.WithdrawCash(aggregateId, amount, replyTo) =>
          state.bankAccount
            .subtract(amount)
            .fold(
              error => {
                replyTo ! WithdrawCashReply.Failed(aggregateId, error)
                Behaviors.same
              },
              { case Result(newBankAccount, event) =>
                effector.persistEvent(event) { _ =>
                  replyTo ! WithdrawCashReply.Succeeded(aggregateId, amount)
                  handleCreated(state.copy(bankAccount = newBankAccount), effector)
                }
              },
            )
      }
    }
}

File: example/src/main/java/com/github/j5ik2o/pekko/persistence/effector/example/javaimpl/persistenceEffector/oop/BankAccountAggregate.java
package com.github.j5ik2o.pekko.persistence.effector.example.javaimpl.persistenceEffector.oop;

import com.github.j5ik2o.pekko.persistence.effector.example.javaimpl.*;
import com.github.j5ik2o.pekko.persistence.effector.javadsl.*;
import org.apache.pekko.actor.typed.Behavior;
import org.apache.pekko.actor.typed.javadsl.Behaviors;

/** Bank Account Aggregate */
public class BankAccountAggregate {
  /**
   * Get actor name
   *
   * @param aggregateId Aggregate ID
   * @return Actor name
   */
  public static String actorName(BankAccountId aggregateId) {
    return aggregateId.getAggregateTypeName() + "-" + aggregateId.asString();
  }

  /**
   * Create a bank account actor
   *
   * @param aggregateId Aggregate ID
   * @return Actor behavior
   */
  public static Behavior<BankAccountCommand> create(BankAccountId aggregateId) {
    return create(aggregateId, PersistenceMode.PERSISTENCE);
  }

  /**
   * Create a bank account actor
   *
   * @param aggregateId Aggregate ID
   * @param persistenceMode Persistence mode
   * @return Actor behavior
   */
  public static Behavior<BankAccountCommand> create(
      BankAccountId aggregateId, PersistenceMode persistenceMode) {
    var config =
        PersistenceEffectorConfig
            .<BankAccountAggregateState, BankAccountEvent, BankAccountCommand>create(
                PersistenceId.ofUniqueId(actorName(aggregateId)),
                new BankAccountAggregateState.NotCreated(aggregateId),
                BankAccountAggregateState::applyEvent)
            .withPersistenceMode(persistenceMode)
            .withSnapshotCriteria(SnapshotCriteria.every(2))
            .withRetentionCriteria(RetentionCriteria.ofSnapshotEvery(2));

    return Behaviors.setup(
        ctx ->
            PersistenceEffector.fromConfig(
                config,
                (initialState, effector) ->
                    switch (initialState) {
                      case BankAccountAggregateState.NotCreated notCreated ->
                          handleNotCreated(notCreated, effector);
                      case BankAccountAggregateState.Created created ->
                          handleCreated(created, effector);
                      default ->
                          throw new IllegalStateException("Unexpected state: " + initialState);
                    }));
  }

  private static Behavior<BankAccountCommand> handleNotCreated(
      BankAccountAggregateState.NotCreated state,
      PersistenceEffector<BankAccountAggregateState, BankAccountEvent, BankAccountCommand>
          effector) {
    return Behaviors.receiveMessage(
        command ->
            switch (command) {
              case BankAccountCommand.Create create -> {
                var result = BankAccount.create(create.getAggregateId());
                yield effector.persistEvent(
                    result.getEvent(),
                    newState -> {
                      create
                          .replyTo()
                          .tell(CommandReply.CreateReply.succeeded(create.getAggregateId()));
                      var created =
                          new BankAccountAggregateState.Created(
                              state.getAggregateId(), result.getState());
                      return handleCreated(created, effector);
                    });
              }
              default -> throw new IllegalStateException("Unexpected command: " + command);
            });
  }

  private static Behavior<BankAccountCommand> handleCreated(
      BankAccountAggregateState.Created state,
      PersistenceEffector<BankAccountAggregateState, BankAccountEvent, BankAccountCommand>
          effector) {
    return effector.persistSnapshot(
        state,
        newState ->
            Behaviors.receiveMessage(
                command ->
                    switch (command) {
                      case BankAccountCommand.Stop stop -> {
                        stop.replyTo()
                            .tell(CommandReply.StopReply.succeeded(stop.getAggregateId()));
                        yield Behaviors.<BankAccountCommand>stopped();
                      }
                      case BankAccountCommand.GetBalance getBalance -> {
                        getBalance
                            .replyTo()
                            .tell(
                                CommandReply.GetBalanceReply.succeeded(
                                    getBalance.getAggregateId(), state.bankAccount().getBalance()));
                        yield Behaviors.<BankAccountCommand>same();
                      }
                      case BankAccountCommand.DepositCash depositCash -> {
                        var result = state.bankAccount().add(depositCash.amount());
                        if (result.isLeft()) {
                          depositCash
                              .replyTo()
                              .tell(
                                  CommandReply.DepositCashReply.failed(
                                      depositCash.getAggregateId(), result.getLeft()));
                          yield Behaviors.<BankAccountCommand>same();
                        } else {
                          var newBankAccount = result.getRight();
                          yield effector.persistEvent(
                              newBankAccount.getEvent(),
                              newState2 -> {
                                depositCash
                                    .replyTo()
                                    .tell(
                                        CommandReply.DepositCashReply.succeeded(
                                            depositCash.getAggregateId(), depositCash.amount()));
                                var created =
                                    new BankAccountAggregateState.Created(
                                        state.getAggregateId(), newBankAccount.getState());
                                return handleCreated(created, effector);
                              });
                        }
                      }
                      case BankAccountCommand.WithdrawCash withdrawCash -> {
                        var result = state.bankAccount().subtract(withdrawCash.amount());
                        if (result.isLeft()) {
                          withdrawCash
                              .replyTo()
                              .tell(
                                  CommandReply.WithdrawCashReply.failed(
                                      withdrawCash.getAggregateId(), result.getLeft()));
                          yield Behaviors.<BankAccountCommand>same();
                        } else {
                          var newBankAccount = result.getRight();
                          yield effector.persistEvent(
                              newBankAccount.getEvent(),
                              newState2 -> {
                                withdrawCash
                                    .replyTo()
                                    .tell(
                                        CommandReply.WithdrawCashReply.succeeded(
                                            withdrawCash.getAggregateId(), withdrawCash.amount()));
                                var created =
                                    new BankAccountAggregateState.Created(
                                        state.getAggregateId(), newBankAccount.getState());
                                return handleCreated(created, effector);
                              });
                        }
                      }
                      default -> throw new IllegalStateException("Unexpected command: " + command);
                    }));
  }
}
